ceno_zkvm/src/instructions/global.rs

```diff
@@ -0,0 +1,817 @@
+    chip_handler::general::PublicIOQuery,
+    e2e::ShardContext,
+    error::ZKVMError,
+    gadgets::{Poseidon2Config, RoundConstants},
+    instructions::riscv::constants::UINT_LIMBS,
+    scheme::septic_curve::{SepticExtension, SepticPoint},
+    structs::{ProgramParams, RAMType},
+    tables::RMMCollections,
+    witness::LkMultiplicity,
+};
+use ff_ext::{ExtensionField, FieldInto, POSEIDON2_BABYBEAR_WIDTH, SmallField};
+use gkr_iop::{
+    chip::Chip, circuit_builder::CircuitBuilder, error::CircuitBuilderError, gkr::layer::Layer,
+    selector::SelectorType, utils::lk_multiplicity::Multiplicity,
+};
+use itertools::{Itertools, chain};
+use multilinear_extensions::{
+    Expression, StructuralWitInType::EqualDistanceSequence, ToExpr, WitIn, util::max_usable_threads,
+};
+use p3::{
+    field::{Field, FieldAlgebra},
+    matrix::dense::RowMajorMatrix,
+    symmetric::Permutation,
+};
+use rayon::{
+    iter::{IndexedParallelIterator, IntoParallelIterator, ParallelExtend, ParallelIterator},
+    prelude::ParallelSliceMut,
+    slice::ParallelSlice,
+};
+use std::ops::Deref;
+use witness::{InstancePaddingStrategy, next_pow2_instance_padding, set_val};
+
+use crate::{
+    instructions::{Instruction, riscv::constants::UInt},
+    scheme::constants::SEPTIC_EXTENSION_DEGREE,
+};
+
+/// A record for a read/write into the global set
+#[derive(Default, Debug, Clone)]
+pub struct GlobalRecord {
+    pub addr: u32,
+    pub ram_type: RAMType,
+    pub value: u32,
+    pub shard: u64,
+    pub local_clk: u64,
+    pub global_clk: u64,
+    pub is_write: bool,
+}
+
+/// An EC point corresponding to a global read/write record
+/// whose x-coordinate is derived from Poseidon2 hash of the record
+#[derive(Clone, Debug)]
+pub struct GlobalPoint<E: ExtensionField> {
+    pub nonce: u32,
+    pub point: SepticPoint<E::BaseField>,
+}
+
+impl GlobalRecord {
+    pub fn to_ec_point<
+        E: ExtensionField,
+        P: Permutation<[E::BaseField; POSEIDON2_BABYBEAR_WIDTH]>,
+    >(
+        &self,
+        hasher: &P,
+    ) -> GlobalPoint<E> {
+        let mut nonce = 0;
+        let mut input = [
+            E::BaseField::from_canonical_u32(self.addr),
+            E::BaseField::from_canonical_u32(self.ram_type as u32),
+            E::BaseField::from_canonical_u32(self.value & 0xFFFF), // lower 16 bits
+            E::BaseField::from_canonical_u32((self.value >> 16) & 0xFFFF), // higher 16 bits
+            E::BaseField::from_canonical_u64(self.shard),
+            E::BaseField::from_canonical_u64(self.global_clk),
+            E::BaseField::from_canonical_u32(nonce),
+            E::BaseField::ZERO,
+            E::BaseField::ZERO,
+            E::BaseField::ZERO,
+            E::BaseField::ZERO,
+            E::BaseField::ZERO,
+            E::BaseField::ZERO,
+            E::BaseField::ZERO,
+            E::BaseField::ZERO,
+            E::BaseField::ZERO,
+        ];
+
+        let prime = E::BaseField::order().to_u64_digits()[0];
+        loop {
```

Comment:
An improvement: we can do parallel search + find_any just like plonk3 grinding api

https://github.com/Plonky3/Plonky3/blob/658115aae28312e97cd3ef38d1ab046839f734a8/challenger/src/grinding_challenger.rs#L56-L62